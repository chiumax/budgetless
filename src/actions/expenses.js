import uuid from "uuid";
import database from "../firebase/firebase";

//  ADD EXPENSE
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

// REMOVE EXPENSE
export const removeExpense = ({ id = "" } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = ({ id = "" } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

// EDIT EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};

//  SET_EXPENSES
export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses`)
      .once("value")
      .then(snapshot => {
        const expenses = [];
        snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
