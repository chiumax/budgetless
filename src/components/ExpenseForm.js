import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";
import swal from "sweetalert";
import NavigationPrompt from "react-router-navigation-prompt";
import { BigNumber } from "bignumber.js";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense
        ? BigNumber(props.expense.amount)
            .div(100)
            .toString(10)
        : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: "",
      onFormDirty: false,
      modalTitle: "Are You Sure?",
      modalText:
        "Once you leave the page, you will not be able to recover your edits.",
      modalIcon: "warning",
      modalButtons: {
        confirm: { text: "Leave", value: true },
        cancel: "Stay"
      }
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    const formdirty = e.target.value.length > 0;
    this.setState(() => ({
      description,
      onFormDirty: formdirty
    }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    const formdirty = e.target.value.length > 0;

    this.setState(() => ({
      note,
      onFormDirty: formdirty
    }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    const formdirty = e.target.value.length > 0;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount, onFormDirty: formdirty }));
    }
  };

  onDateChange = createdAt => {
    const formdirty = true;

    if (createdAt) {
      this.setState(() => ({ createdAt, onFormDirty: formdirty }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      calendarFocused: focused
    }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      //   Set error state equal to 'Please provide description and amount.'
      this.setState(() => ({
        error: "Please provide a description and an amount value."
      }));
    } else {
      //   Clear the error
      const modalTemp = this.props.expense
        ? "Expense Edited!"
        : "Expense Created!";
      this.setState(() => ({
        error: "",
        onFormDirty: false,
        modalTitle: modalTemp,
        modalText:
          "You can edit/view expenses by clicking on them at the Dashboard",
        modalIcon: "success",
        modalButtons: {
          confirm: { text: "Ok", value: true }
        }
      }));
      if (this.props.expenseLength === true) {
        this.setState(() => ({
          modalIcon: "error",
          modalButtons: {
            confirm: { text: "Ok", value: true }
          },
          modalText:
            "We're sorry for limiting this account to 5 expenses since we are currently hosted using a free service. ",
          modalTitle: "We're sorry."
        }));
      }
      this.props.onSubmit({
        description: this.state.description,
        amount: BigNumber(this.state.amount)
          .multipliedBy(100)
          .toString(10),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  componentWillUnmount() {}
  hideModal = () => {
    swal.close();
  };
  render() {
    return (
      <div>
        <NavigationPrompt when={this.state.onFormDirty}>
          {({ isActive, onConfirm, onCancel }) => {
            if (isActive) {
              swal({
                title: this.state.modalTitle,
                text: this.state.modalText,
                icon: this.state.modalIcon,
                buttons: this.state.modalButtons,
                dangerMode: true,
                closeOnEsc: false,
                closeOnClickOutside: false
              }).then(willDelete => {
                if (willDelete) {
                  onConfirm();
                } else {
                  onCancel();
                  this.hideModal();
                }
              });
            }
          }}
        </NavigationPrompt>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
            maxLength="50"
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
            maxLength="20"
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            value={this.state.note}
            onChange={this.onNoteChange}
            placeholder="Add a note for your expense (optional)"
            maxLength="500"
          />
          <button>{this.props.expense ? "Edit Expense" : "Add Expense"}</button>
        </form>
      </div>
    );
  }
}
