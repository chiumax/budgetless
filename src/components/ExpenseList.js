import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = props => (
  <div>
    Some text to make this easy to identify on the screen.
    {props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      props.expenses.map(item => {
        return <ExpenseListItem key={item.id} {...item} />;
      })
    )}
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
