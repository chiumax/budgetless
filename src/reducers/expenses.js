// EXPENSES REDUCER
const expensesReducerDefaultState = [];
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
      break;
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => {
        return id != action.id;
      });
      break;
    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        } else {
          return expense;
        }
      });

    case "SET_EXPENSES":
      return action.expenses;
      break;
    default:
      return state;
      break;
  }
};

export default expensesReducer;
