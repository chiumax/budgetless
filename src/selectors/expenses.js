import moment from "moment";
import { BigNumber } from "bignumber.js";

// GET VISIBLE EXPENSES
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());
      return textMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
      if (sortBy === "amount") {
        return BigNumber(a.amount).lt(b.amount) ? 1 : -1;
      }
    });
};

export default getVisibleExpenses;
