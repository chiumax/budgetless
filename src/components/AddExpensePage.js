import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";
import Anime from "react-anime";

export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    if (this.props.expenseLength === false) {
      this.props.startAddExpense(expense);
    }
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <Anime
          easing="easeInExpo"
          duration={1000}
          direction="normal"
          translateX={["-13rem", 0]}
          opacity={[0, 1]}
        >
          <div className="page-header">
            <div className="content-container">
              <h1 className="page-header__title">Add An Expense</h1>
            </div>
          </div>
          <div className="content-container">
            <ExpenseForm
              onSubmit={this.onSubmit}
              expenseLength={this.props.expenseLength}
            />
          </div>
        </Anime>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    startAddExpense: expense => dispatch(startAddExpense(expense))
  };
};

const mapStateToProps = state => {
  return {
    expenseLength: state.expenses.length >= 5
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddExpensePage);
