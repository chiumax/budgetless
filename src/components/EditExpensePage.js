import React from "react";
import { connect } from "react-redux";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";
import Anime from "react-anime";

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    if (this.props.expenseLength === false) {
      this.props.startEditExpense(this.props.expense.id, expense);
    }

    this.props.history.push("/");
  };
  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <Anime
          easing="easeInExpo"
          duration={1000}
          direction="normal"
          translateX={["-13rem", 0]}
          opacity={[0, 1]}
        >
          <div className="page-header">
            <div className="content-container">
              <h1 className="page-header__title">Edit Expense</h1>
            </div>
          </div>
          <div className="content-container">
            <ExpenseForm
              expense={this.props.expense}
              onSubmit={this.onSubmit}
              expenseLength={false}
            />
            <button className="button button--secondary" onClick={this.onClick}>
              Remove Expense
            </button>
          </div>
        </Anime>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => {
      return expense.id === props.match.params.id;
    })
  };
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: id => dispatch(startRemoveExpense(id))
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
