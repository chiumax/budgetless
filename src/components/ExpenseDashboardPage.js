import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";
import Anime from "react-anime";

const ExpenseDashboardPage = () => (
  <div>
    <Anime
      easing="easeInExpo"
      duration={1000}
      direction="normal"
      delay={(el, index) => index * 240}
      translateX={["-13rem", 0]}
      opacity={[0, 1]}
    >
      <div>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />
      </div>
    </Anime>
  </div>
);

export default ExpenseDashboardPage;
