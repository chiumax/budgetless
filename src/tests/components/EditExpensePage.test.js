import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let editExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      startRemoveExpense={startRemoveExpense}
      history={history}
      expense={expenses[1]}
    />
  );
});

test("should render EditExpensePage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle editExpense", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

test("should handle startRemoveExpense", () => {
  wrapper.find("button").prop("onClick")(expenses[1].id);
  expect(startRemoveExpense).toHaveBeenLastCalledWith({
    id: expenses[1].id
  });
});
