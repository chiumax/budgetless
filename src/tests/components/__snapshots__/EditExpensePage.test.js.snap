// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": "1234",
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
